File: sdk/examples/20_send_compliment.rs
use api_types::api::transactions::ApiTxStatus;
use etopay_sdk::types::{currencies::CryptoAmount, newtypes::AccessToken};
use rust_decimal_macros::dec;
use std::time::Duration;
use testing::USER_HANS34;
mod utils;
use tokio::time;
use utils::init_sdk;

#[allow(clippy::unwrap_used)]
#[tokio::main]
async fn main() {
    // Initialize SDK for sender, create new user and migrate wallet
    let (mut sdk, _cleanup) = init_sdk().await;
    let user: utils::TestUser = (*USER_HANS34).clone().into();

    // the `init_sdk()` function generates an access token for `satoshi`.
    // in this example we use `hans34` user. Therefore, we generate a new access token for the `hans34` user.
    let access_token = testing::get_access_token(&user.username, user.password.as_str())
        .await
        .access_token;
    let access_token = AccessToken::try_from(access_token).unwrap();
    sdk.refresh_access_token(Some(access_token)).await.unwrap();

    // Create new user
    sdk.create_new_user(&user.username).await.unwrap();
    sdk.init_user(&user.username).await.unwrap();

    // Create / init new wallet from mnemonic
    sdk.set_wallet_password(&user.pin, &user.password).await.unwrap();
    sdk.create_wallet_from_existing_mnemonic(&user.pin, &user.mnemonic)
        .await
        .unwrap();

    // Fetch networks from backend
    let networks = sdk.get_networks().await.unwrap();
    let iota_network_key = &networks.first().unwrap().key;
    sdk.set_network(iota_network_key.to_string()).await.unwrap();

    // Generate address and get balance
    let address = sdk.generate_new_address(&user.pin).await.unwrap(); // this is needed, otherwise the balance will be 0 and tx will fail
    let balance = sdk.get_balance(&user.pin).await.unwrap();
    println!("Balance: {:#?} on address {}", balance, address);

    // Create purchase request
    let product_hash = "2cf24dba5fb0a30e26e83b2ac5b9e29e1b161e5c1fa7425e73043362938b9824";
    let app_data = "{\"imageUrl\":\"https://httpbin.org/\",\"imageId\":\"a846ad10-fc69-4b22-b442-5dd740ace361\"}";
    let purchase_type = "CLIK";

    let amount = CryptoAmount::try_from(dec!(2.0)).unwrap();
    let purchase_id = sdk
        .create_purchase_request("alice", amount, product_hash, app_data, purchase_type)
        .await
        .unwrap();
    println!("Purchase_id {} ", purchase_id); // print the purchase id to facilitate debugging

    // Wait 3 min while tx status becomes valid
    let result = time::timeout(Duration::from_secs(3 * 60), async {
        loop {
            time::sleep(Duration::from_secs(5)).await;
            let details = sdk.get_purchase_details(&purchase_id).await.unwrap();
            match details.status {
                ApiTxStatus::Valid => {
                    println!("Purchase request valid, moving on...");
                    break;
                }
                ApiTxStatus::Invalid(r) => {
                    panic!("Purchase request invalid! Reason: {:?}. Exiting", r);
                }
                ApiTxStatus::WaitingForVerification(r) => {
                    panic!("Purchase request waiting for verification! Reason: {:?}.", r);
                }
                _ => {}
            }
        }
    })
    .await;
    if result.is_err() {
        panic!("Timeout reached while waiting for purchase request to become valid");
    }

    // Step 4: Confirm purchase request (perform actual wallet transaction)
    sdk.confirm_purchase_request(&user.pin, &purchase_id).await.unwrap();

    // Wait 3 min while tx status becomes completed
    let result = time::timeout(Duration::from_secs(3 * 60), async {
        loop {
            time::sleep(Duration::from_secs(5)).await;
            let status = sdk.get_purchase_details(&purchase_id).await.unwrap().status;
            println!(" - Status: {:?}", status);
            if status == ApiTxStatus::Completed {
                println!("Purchase request completed, done!");
                break;
            } else if status == ApiTxStatus::Failed {
                panic!("Purchase request failed");
            }
        }
    })
    .await;
    if result.is_err() {
        panic!("Timeout reached while waiting for purchase request to complete");
    }

    // Check new balance
    let balance = sdk.get_balance(&user.pin).await.unwrap();
    println!("New Balance: {:#?}", balance);
}

File: bindings/wasm/examples/webapp/examples/20-send_compliment.js
import * as wasm from "etopay-sdk-wasm";
import { sleep, initSdk, PIN } from "./util";

async function main() {
    console.log("start");
    const sdk = await initSdk();

    let username = "alice";
    let password = "correcthorsebatterystaple";

    let mnemonic = process.env.MNEMONIC;

    // ===========================================================================
    // Step 1: Initialize SDK for sender, create new user and wallet
    // ===========================================================================

    await sdk.createNewUser(username);
    await sdk.initializeUser(username);
    await sdk.setWalletPassword(PIN, password);
    await sdk.initWalletFromMnemonic(PIN, mnemonic);

    // fetch networks from backend
    let networks = await sdk.getNetworks();
    // set the network configuration for the wallet
    sdk.setNetwork(networks[0].key);

    let address = await sdk.generateNewAddress(PIN);
    console.log("Address:", address);

    let balance = await sdk.getWalletBalance(PIN);
    console.log("Balance:", balance);

    // ===========================================================================
    // Step 3: Create purchase request
    // ===========================================================================

    let product_hash = "2cf24dba5fb0a30e26e83b2ac5b9e29e1b161e5c1fa7425e73043362938b9824";
    let app_data = "{\"imageUrl\":\"https://httpbin.org/\",\"imageId\":\"a846ad10-fc69-4b22-b442-5dd740ace361\"}";
    let purchase_type = "CLIK";

    let purchase_id = await sdk.createPurchaseRequest("satoshi", 2, product_hash, app_data, purchase_type);
    console.log("Purchase Request created: ", purchase_id);

    console.log("Waiting for purchase request to be valid:")
    while (true) {
        await sleep(5000);

        let details = await sdk.getPurchaseDetails(purchase_id);
        console.log("Details: ", details.toJSON());

        if (details.status == wasm.TxStatus.Valid) {
            console.log("Purchase request valid, moving on...");
            break;
        } else if (details.status == wasm.TxStatus.Invalid) {
            console.log("Purchase request invalid, exiting!");
            return
        }

    }


    // ===========================================================================
    // Step 4: Confirm purchase request (perform actual wallet transaction)
    // ===========================================================================

    console.log("Confirming purchase request...");
    await sdk.confirmPurchaseRequest(PIN, purchase_id);
    console.log("Confirming purchase request done, monitoring the status for 60 seconds or until it is completed:");
    for (let i = 0; i < 6; i++) {
        await sleep(10000);

        let details = await sdk.getPurchaseDetails(purchase_id);
        console.log("Status: ", details.toJSON());

        if (details.status == wasm.TxStatus.Completed) {
            console.log("Purchase request completed, done!");
            break;
        }
    }
    let new_balance = await sdk.getWalletBalance(PIN);
    console.log("New Balance:", new_balance);

}

export { main }
